lib/membrane/core/child/pad_model.ex

```diff
@@ -91,29 +86,57 @@ defmodule Membrane.Core.Child.PadModel do
     |> Map.new()
   end
 
+  @spec get_data(Child.state_t(), Pad.ref_t()) :: {:ok, Pad.Data.t() | any} | unknown_pad_error_t
+  def get_data(%{pads: %{data: data}}, pad_ref)
+      when is_map_key(data, pad_ref) do
+    data
+    |> Map.get(pad_ref)
+    ~> {:ok, &1}
+  end
+
+  def get_data(_state, pad_ref), do: {:error, {:unknown_pad, pad_ref}}
+
   @spec get_data(Child.state_t(), Pad.ref_t(), keys :: atom | [atom]) ::
           {:ok, Pad.Data.t() | any} | unknown_pad_error_t
-  def get_data(state, pad_ref, keys \\ []) do
-    with :ok <- assert_instance(state, pad_ref) do
-      state
-      |> Bunch.Access.get_in(data_keys(pad_ref, keys))
-      ~> {:ok, &1}
-    end
+  def get_data(%{pads: %{data: data}}, pad_ref, keys)
+      when is_map_key(data, pad_ref) and is_list(keys) do
+    data
+    |> get_in([pad_ref | keys])
+    ~> {:ok, &1}
+  end
+
+  def get_data(%{pads: %{data: data}}, pad_ref, key)
+      when is_map_key(data, pad_ref) and is_atom(key) do
+    data
+    |> get_in([pad_ref, key])
+    ~> {:ok, &1}
+  end
+
+  def get_data(_state, pad_ref, _keys), do: {:error, {:unknown_pad, pad_ref}}
+
+  @spec get_data!(Child.state_t(), Pad.ref_t()) :: Pad.Data.t() | any
+  def get_data!(state, pad_ref) do
+    {:ok, pad_data} = get_data(state, pad_ref)
+    pad_data
   end
 
   @spec get_data!(Child.state_t(), Pad.ref_t(), keys :: atom | [atom]) :: Pad.Data.t() | any
-  def get_data!(state, pad_ref, keys \\ []) do
+  def get_data!(state, pad_ref, keys) do
     {:ok, pad_data} = get_data(state, pad_ref, keys)
     pad_data
   end
 
   @spec set_data(Child.state_t(), Pad.ref_t(), keys :: atom | [atom], value :: term()) ::
           Type.stateful_t(:ok | unknown_pad_error_t, Child.state_t())
   def set_data(state, pad_ref, keys \\ [], value) do
-    with {:ok, state} <- {assert_instance(state, pad_ref), state} do
-      state
-      |> Bunch.Access.put_in(data_keys(pad_ref, keys), value)
-      ~> {:ok, &1}
+    case assert_instance(state, pad_ref) do
+      :ok ->
+        state
+        |> Bunch.Access.put_in(data_keys(pad_ref, keys), value)
```

Comment:
```suggestion
        put_in(data_keys(pad_ref, keys), value)
```
Bunch.Access seems not to help there and it's an additional function call